File: GalaxyMonkey/SpriteCatalog.swift
//
//  SpriteCatalog.swift
//  GalaxyMonkey
//
//  Single entry point for loading game art from Assets.xcassets.
//  Adding a sprite = one enum case + one .imageset in Assets.xcassets.
//
//  All textures are optional: art is generated incrementally via the
//  SpriteCook pipeline. Subsystems should check `hasTexture(...)` and
//  fall back to primitive shape nodes when art is missing.
//

import SpriteKit
import UIKit

enum Sprite: String, CaseIterable {
    case player        = "Player"
    case bullet        = "Bullet"
    case enemyBullet   = "EnemyBullet"
    case bomb          = "Bomb"
    case goldenBanana  = "GoldenBanana"
    case preppyLeft    = "PreppyLeft"
    case preppyRight   = "PreppyRight"
    case whiteLeft     = "WhiteLeft"
    case whiteRight    = "WhiteRight"
    case shadyLeft     = "ShadyLeft"
    case shadyRight    = "ShadyRight"
    case gorilla       = "Gorilla"
    case gorillaWindup = "GorillaWindup"
    case droneSwarm    = "DroneSwarm"
    case heavyCosmonautLeft  = "HeavyCosmonautLeft"
    case heavyCosmonautRight = "HeavyCosmonautRight"
    case plasmaJelly   = "PlasmaJelly"
    case astroSniperLeft  = "AstroSniperLeft"
    case astroSniperRight = "AstroSniperRight"
    case miniBossLeft  = "MiniBossLeft"
    case miniBossRight = "MiniBossRight"
    case spaceBackdrop = "SpaceBackdrop"
    case explosion     = "Explosion"
    case title         = "Title"
    case pauseIcon     = "PauseIcon"
    case lifeHeart     = "LifeHeart"
    // Parallax backdrop set + glow halo.
    case sun           = "Sun"
    case mercury       = "Mercury"
    case venus         = "Venus"
    case earth         = "Earth"
    case mars          = "Mars"
    case jupiter       = "Jupiter"
    case saturn        = "Saturn"
    case uranus        = "Uranus"
    case neptune       = "Neptune"
    case glow          = "Glow"
}

enum SpriteCatalog {

    private static var cache: [Sprite: SKTexture] = [:]
    private static var missing: Set<Sprite> = []

    /// Returns nil if the imageset is missing from Assets.xcassets. Callers
    /// should branch on nil and draw a placeholder primitive instead.
    static func texture(for sprite: Sprite) -> SKTexture? {
        if let cached = cache[sprite] { return cached }
        if missing.contains(sprite) { return nil }
        guard UIImage(named: sprite.rawValue) != nil else {
            missing.insert(sprite)
            return nil
        }
        let texture = SKTexture(imageNamed: sprite.rawValue)
        texture.filteringMode = .linear
        cache[sprite] = texture
        return texture
    }

    static func hasTexture(_ sprite: Sprite) -> Bool {
        texture(for: sprite) != nil
    }

    static func preload() {
        Sprite.allCases.forEach { _ = texture(for: $0) }
    }
}

File: android/core/src/main/kotlin/dev/copt/galaxymonkey/SpriteCatalog.kt
package dev.copt.galaxymonkey

import com.badlogic.gdx.graphics.g2d.TextureAtlas
import com.badlogic.gdx.graphics.g2d.TextureRegion

// All textures are optional: art is generated incrementally via the
// SpriteCook pipeline. Subsystems check has() and fall back to
// primitive shape drawing when art is missing.

enum class Sprite(val regionName: String) {
    PLAYER("Player"),
    BULLET("Bullet"),
    ENEMY_BULLET("EnemyBullet"),
    BOMB("Bomb"),
    GOLDEN_BANANA("GoldenBanana"),
    PREPPY_LEFT("PreppyLeft"),
    PREPPY_RIGHT("PreppyRight"),
    WHITE_LEFT("WhiteLeft"),
    WHITE_RIGHT("WhiteRight"),
    SHADY_LEFT("ShadyLeft"),
    SHADY_RIGHT("ShadyRight"),
    GORILLA_LEFT("GorillaLeft"),
    GORILLA_RIGHT("GorillaRight"),
    GORILLA_WINDUP("GorillaWindup"),
    DRONE_SWARM_LEFT("DroneSwarmLeft"),
    DRONE_SWARM_RIGHT("DroneSwarmRight"),
    HEAVY_COSMONAUT_LEFT("HeavyCosmonautLeft"),
    HEAVY_COSMONAUT_RIGHT("HeavyCosmonautRight"),
    PLASMA_JELLY("PlasmaJelly"),
    ASTRO_SNIPER_LEFT("AstroSniperLeft"),
    ASTRO_SNIPER_RIGHT("AstroSniperRight"),
    MINI_BOSS_LEFT("MiniBossLeft"),
    MINI_BOSS_RIGHT("MiniBossRight"),
    SPACE_BACKDROP("SpaceBackdrop"),
    EXPLOSION("Explosion"),
    TITLE("Title"),
    PAUSE_ICON("PauseIcon"),
    LIFE_HEART("LifeHeart"),
    SUN("Sun"),
    MERCURY("Mercury"),
    VENUS("Venus"),
    EARTH("Earth"),
    MARS("Mars"),
    JUPITER("Jupiter"),
    SATURN("Saturn"),
    URANUS("Uranus"),
    NEPTUNE("Neptune"),
    GLOW("Glow"),
    LUNA("Luna"),
    IO("Io"),
    EUROPA("Europa"),
    GANYMEDE("Ganymede"),
    CALLISTO("Callisto"),
    TITAN("Titan"),
    ROCK_01("Rock01"),
    ROCK_02("Rock02"),
    ROCK_03("Rock03"),
    ROCK_04("Rock04"),
}

object SpriteCatalog {

    private lateinit var atlas: TextureAtlas
    private val cache = mutableMapOf<Sprite, TextureRegion>()
    private val missing = mutableSetOf<Sprite>()

    fun init(atlas: TextureAtlas) {
        this.atlas = atlas
        cache.clear()
        missing.clear()
    }

    fun region(sprite: Sprite): TextureRegion? {
        if (!::atlas.isInitialized) return null
        cache[sprite]?.let { return it }
        if (sprite in missing) return null
        val region = atlas.findRegion(sprite.regionName)
        if (region == null) {
            missing.add(sprite)
            return null
        }
        cache[sprite] = region
        return region
    }

    fun has(sprite: Sprite): Boolean = region(sprite) != null

    fun preload() {
        Sprite.entries.forEach { region(it) }
    }
}
